### ISSUE TITLE
external workers cannot handle big classpath on windows

### ISSUE THREAD

**Issue by TwoStone** [gwtproject/gwt#9530]
**GWT version: 2.7, 2.8, and earlier** 
**Browser (with version): n/a** 
**Operating System: Windows** 

---

##### Description

On windows the lenght of commands passed to cmd is limited to 8192 characters.
When calling the gwt compiler with a big classpath on the commandline will crash with some error telling you that your commandline is too long.
To work around this problem it is a common practice to set the classpath via the "CLASSPATH"-environment variable as it can handle quite more characters.

unfortunatly, when compiling more the one permutation in forked mode, the compiler tries to build up a the classpath for the forked process and passes this via the process arguments.
(see ExternalPermutationWorker#launchExternalWorker(TreeLogger, int))

##### Steps to reproduce

1. setup a gwt project with a big classpath
2. configure 2(+) permutations (i. e. safari, firefox)
3. run gwt-compiler on windows with 2(+) workers

**Comment by malafoss** [comment]
```
diff --git a/dev/core/src/com/google/gwt/dev/ExternalPermutationWorkerFactory.java b/dev/core/src/com/google/gwt/dev/ExternalPermutationWorkerFactory.java
index b6d4573fd..0cde94ed6 100644
--- a/dev/core/src/com/google/gwt/dev/ExternalPermutationWorkerFactory.java
+++ b/dev/core/src/com/google/gwt/dev/ExternalPermutationWorkerFactory.java
@@ -42,6 +42,7 @@ import java.util.Collections;
 import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.Random;
 import java.util.Set;

@@ -260,12 +261,10 @@ public class ExternalPermutationWorkerFactory extends PermutationWorkerFactory {
     random.nextBytes(cookieBytes);
     String cookie = StringUtils.toHexString(cookieBytes);

-    // Cook up the classpath, main class, and extra args
-    args.addAll(Arrays.asList("-classpath",
-        ManagementFactory.getRuntimeMXBean().getClassPath(),
-        CompilePermsServer.class.getName(), "-host", "localhost", "-port",
-        String.valueOf(port), "-logLevel", logLevel.toString(), "-cookie",
-        cookie));
+    // Cook up the main class, and extra args
+    args.addAll(Arrays.asList(CompilePermsServer.class.getName(),
+        "-host", "localhost", "-port", String.valueOf(port),
+        "-logLevel", logLevel.toString(), "-cookie", cookie));

     // Filter undesirable arguments
     for (Iterator<String> iter = args.iterator(); iter.hasNext();) {
@@ -278,6 +277,10 @@ public class ExternalPermutationWorkerFactory extends PermutationWorkerFactory {
     ProcessBuilder builder = new ProcessBuilder();
     builder.command(args);

+    // Setup the classpath
+    Map<String, String> env = builder.environment();
+    env.put("CLASSPATH", ManagementFactory.getRuntimeMXBean().getClassPath());
+
     try {
       final Process proc = builder.start();
       final BufferedReader bin = new BufferedReader(new InputStreamReader(
```

**Comment by niloc132** [comment]
@malafoss can you submit this at https://gwt-review.googlesource.com/ so you can sign off on contributing this code to GWT and get it reviewed through our normal channels?

**Comment by malafoss** [comment]
Sure, I can try to do that.

**Comment by malafoss** [comment]
https://gwt-review.googlesource.com/c/gwt/+/21480

**Comment by niloc132** [comment]
21480 never was merged, https://gwt-review.googlesource.com/c/gwt/+/22700 also has the same change, but fewer style nits? 

I've reconfirmed the fix by running the code and getting a good result, then deliberately breaking the classpath env var name and verifying that it no longer worked, and @rdeangelis83 verified that this improves the experience on windows.

**Comment by niloc132** [comment]
Fixed in 2.10.0.
